Add new tutorial (quick reference for adding modules to pipelines) (#1410)
nf-core/nf-co.re: markdown/developers/tutorials/adding_modules_to_pipelines.md

---
title: 'Tutorial: Adding Modules to Pipelines'
subtitle: Quick reference of the steps needed to add a new module to a pipeline.
---

This tutorial is a quick reference overview of the steps you will normally take when adding a new module to a pipeline.

Different pipelines may have different workflows, but the following steps will cover the most common aspects.

1. Install module to the pipeline

   ```bash
   nf-core modules install <tool>/<subtool>
   ```

> 2. Add `include` statement at the top of the pipeline's (sub)workflow to import module

Reviewer: Not sure if the include part would work as is in a subworkflow given the relative path

Author: But you always have to have an include statement no?

https://github.com/nf-core/modules/blob/master/subworkflows/nf-core/bam_markduplicates_picard/main.nf

Reviewer 2: Could mention that a suggestion for this comes from the install command.

4. Create a section in `conf/modules.conf` for the module (with a default results directory and output file pattern)
5. Add any necessary parameters for the module with defaults to `nextflow.config`
6. Insert the parameters into `ext.args` of corresponding `conf/modules.config` (and in some cases in the module reference in the workflow script)
7. Update the `nextflow_schema.json` to include new parameters with

   ```bash
   nf-core schema build
   ```

8. Update `assets/multiqc_config.yml` to include any new MultiQC modules (if any exist) and specify order in the report
9. Add a citation for the new tool/module to `citations.md`
10. Update `docs/USAGE.md` to describe any important information about running of the module (this can be optional in some cases)
11. Update `docs/OUTPUT.md` to describe the directories output files of the module
12. Update `README.md` mentioning the tool is used (optional)
13. Run prettier formatting on the whole repository (if installed)

    ```bash
    prettier -w .
    ```

    - Or run VSCode 'Format Document' function on: `citations.md`, `usage.md`, `output.md`, `README.md`, `modules.json`, `nextflow_schema.json`

>     - If you forget this step you can also post a comment on the open PR once made with `@nf-core bot fix linting`

Author: ```suggestion
    - If you forget this step you can also post a comment on the open PR once made with `@nf-core-bot fix linting`
```

> 

Reviewer 3: should it not be `@nf-core-bot fix linting` with a dash?

Reviewer 4: Judging from the contents of the `.github/workflows` it might be correct https://github.com/nf-core/nf-co.re/blob/a0630dafeffa7262aad2616266478405265e953f/.github/workflows/fix-linting.yml#L11  🤔

Author: Correct!

---
title: 'Tutorial: Adding Modules to Pipelines'
subtitle: Quick reference of the steps needed to add a new module to a pipeline.
---

This tutorial is a quick reference overview of the steps you will normally take when adding a new module to a pipeline.

Different pipelines may have different workflows, but the following steps will cover the most common aspects.

1. Install module to the pipeline

   ```bash
   nf-core modules install <tool>/<subtool>
   ```

> 2. Add `include` statement at the top of the pipeline's (sub)workflow to import module

Author: ```suggestion
2. Add `include` statement at the top of the pipeline's (sub)workflow to import module (as suggested by `nf-core modules install`)
```

3. Insert module execution into the relevant place (sub)workflow script `TOOL(ch_input)`
   - Make sure to mix in the module's version into the version channel, e.g. `ch_versions = ch_versions.mix(TOOL.out.versions)`
   - Make sure to mix any output files for MultiQC into a relevant channel, e.g. `ch_multiqc_files = ch_multiqc.mix(TOOL.out.log)`
4. Create a section in `conf/modules.conf` for the module (with a default results directory and output file pattern)
5. Add any necessary parameters for the module with defaults to `nextflow.config`

> 6. Insert the parameters into `ext.args` of corresponding `conf/modules.config` (and in some cases in the module reference in the workflow script)

Reviewer 5: Can you clarify this please?

```
and in some cases in the module reference in the workflow script)
```

Author: Will do, for reference: `val` inputs to modules e.g. fastp:https://github.com/nf-core/modules/blob/0c078028e29d84f75853552af414019fc5496bab/modules/nf-core/fastp/main.nf#L14

> 6. Insert the parameters into `ext.args` of corresponding `conf/modules.config` (and in some cases in the module reference in the workflow script)

Author: ```suggestion
6. Insert any pipeline level parameters (`params.*`) into `ext.args` of corresponding `conf/modules.config`
    - In some cases these may need to be passed directly to the module itself, e.g. `FASTP( reads, params.save_trimmed_fail,  params.save_merged)`
```

---
title: 'Tutorial: Adding Modules to Pipelines'
subtitle: Quick reference of the steps needed to add a new module to a pipeline.
---

This tutorial is a quick reference overview of the steps you will normally take when adding a new module to a pipeline.

Different pipelines may have different workflows, but the following steps will cover the most common aspects.

1. Install module to the pipeline

>    ```bash

Reviewer: ```suggestion
   ```console
```

Reviewer 6: Disagree with this. If it is not a specific console output `bash` should be used from now on. We added some more code highlighting for it (nf-core is recognized as a function for example), and an automatic copy-button appears next to `bash` code blocks. 
Same for all the other suggested changes.

Reviewer: I see, so console for all console output, but bash for the rest

Reviewer 6: now I am wondering where would be a good place to document this?

Reviewer: In the developers guidelines

Author: HA I WAS RIGHT! IN YO' FACE @maxulysse :-*

Reviewer: I'm pretty sure you have some other `bash` -> `console` markdown stuff to update in your other PR then.
Just so you know ;-)

   nf-core modules install <tool>/<subtool>
   ```

2. Add `include` statement at the top of the pipeline's (sub)workflow to import module (as suggested by `nf-core modules install`)
3. Insert module execution into the relevant place (sub)workflow script `TOOL(ch_input)`
   - Make sure to mix in the module's version into the version channel, e.g. `ch_versions = ch_versions.mix(TOOL.out.versions)`
   - Make sure to mix any output files for MultiQC into a relevant channel, e.g. `ch_multiqc_files = ch_multiqc.mix(TOOL.out.log)`
4. Create a section in `conf/modules.conf` for the module (with a default results directory and output file pattern)
5. Add any necessary parameters for the module with defaults to `nextflow.config`
6. Insert any pipeline level parameters (`params.*`) into the `ext.args` of corresponding `conf/modules.config`
   - In some cases these may need to be passed directly to the module itself, e.g. `FASTP( reads, params.save_trimmed_fail, params.save_merged)`
7. Update the `nextflow_schema.json` to include new parameters with

>    ```bash

Reviewer: ```suggestion
   ```console
```

   nf-core schema build
   ```

8. Update `assets/multiqc_config.yml` to include any new MultiQC modules (if any exist) and specify order in the report
9. Add a citation for the new tool/module to `citations.md`
10. Update `docs/USAGE.md` to describe any important information about running of the module (this can be optional in some cases)
11. Update `docs/OUTPUT.md` to describe the directories output files of the module

> 12. Update `README.md` mentioning the tool is used (optional)

Reviewer 7: ```suggestion
12. Update `README.md` and any pipeline diagrams mentioning the tool is used (optional)
```

13. Run prettier formatting on the whole repository (if installed)

>     ```bash

Reviewer: ```suggestion
    ```console
```

    prettier -w .
    ```

    - Or run VSCode 'Format Document' function on: `citations.md`, `usage.md`, `output.md`, `README.md`, `modules.json`, `nextflow_schema.json`
    - If you forget this step you can also post a comment on the open PR once made with `@nf-core-bot fix linting`

14. Lint the new code with

>     ```bash

Reviewer: ```suggestion
    ```console
```
